### ISSUE TITLE
Cannot transform a `TensorXf` object with shape=(3,) into a `Point3f` object

### ISSUE THREAD

**Issue by FYRichie** [mitsuba-renderer/mitsuba3#280]
## Summary

Cannot transform a `TensorXf` object with shape=(3,) into a `Point3f` object.

## System configuration

System information:

  OS: 11.1
  CPU: b'Intel(R) Core(TM) i7-8750H CPU @ 2.20GHz'
  Python: 3.9.13 (main, Aug 25 2022, 18:29:29) [Clang 12.0.0 ]
  LLVM: 0.0.0

  Dr.Jit: 0.2.2
  Mitsuba: 3.0.2
     Is custom build? False
     Compiled with: AppleClang 13.0.0.13000029
     Variants:
        scalar_rgb
        scalar_spectral
        llvm_ad_rgb

## Description

Hi, I'm using Mitsuba 3 with pytorch to perform inverse rendering, and currently, I've encountered a problem when using the `drjit` function decorator:
```python
@dr.wrap_ad(source="torch", target="drjit")
def add_light(idx: int, intensity: torch.Tensor, position: torch.Tensor, direction: torch.Tensor, divergence_angle: torch.Tensor) -> None:
    scene_dict["SpotLight%d" % idx] = {
        "type": "spot",
        "intensity": {
            "type": "rgb",
            "value": intensity,
        },
        "to_world": ST.look_at(
            origin=mi.Point3f(position),
            target=mi.Point3f(direction),
            up=[0, 0, 1],
        ),
        "cutoff_angle": divergence_angle,
        "beam_width": 16.124722,  # TODO: prediction?
    }
```
The decorator turns input `torch.Tensor` objects into `mitsuba.TensorXf` objects. However, the `ScalarTransform4f.look_at()` functions must take `origin`, `target`, and `up` as `mitsuba.Point3f`. I cannot find any functions in the official document that turns `TensorXf` with shape (3,) into `Point3f`. I want to remain the transformation process to be differentiable. Is there any way to fulfill this requirement? Thanks.

**Comment by FYRichie** [comment]
Btw, if I turn a `TensorXf` object into a `Point3f` or `Array3f` object, will it loss its gradients?

**Comment by Speierers** [comment]
@FYRichie in you example you are creating the scene within the wrapped function. This is not going to work as the scene loading in Mitsuba 3 is always performed using "scalar" inputs. The right way to achieve this is to load the scene once outside of the wrapped function and then only assign the new parameters using `params = mi.traverse(scene)` as done in the tutorial.

**Comment by FYRichie** [comment]
@Speierers Thank you for your reply, but in the documentation, it shows that only those parameters shown in the `params` can be changed. For example, if we have a `SpotLight`:
```python
scene: mi.Scene = mi.load_dict(
    {
        "type": "scene",
        # The keys below correspond to object IDs and can be chosen arbitrarily
        "integrator": {"type": "path"},
        # "light": {"type": "constant"},
        "SpotLight": {
            "type": "spot",
            "to_world": ST.look_at(origin=[1, 1, 1], target=[1, 2, 1], up=[0, 0, 1]),
            "intensity": {
                "type": "rgb",
                "value": [0.1, 0.2, 0.3],
            },
            "cutoff_angle": 30.0,
            "beam_width": 15,
        },
    }
)
print(scene_params)
```
The output will look like this:

| Name | Flags | Type | Parent |
| --- | --- | --- | --- | 
| SpotLight.intensity.value | ∂ | Color3f | SRGBReflectanceSpectrum |
| SpotLight.texture.value | ∂ | float | UniformSpectrum |
| SpotLight.to_world |  | Transform4f | SpotLight |

In my task, I wish to update all the parameters of the `SpotLight`, including `cutoff_angle`, `bean_width`, etc...
Are there any solutions?

**Comment by FYRichie** [comment]
Also, if I wish to add a new light source into the current scene params, it will cause a `KeyError`.

**Comment by Speierers** [comment]
Adding new objects to the scene is currently not supported. If you really need to do this in your wrapped function, they I would suggest you convert those non-scalar values to scalar values using `dr.slice()` before passing them to the scene dictionary. This will be pretty inefficient and you will lose the differentiability for those parameters.

Regarding the missing parameters on the `SpotLight` plugin, those are currently not exposed as they might introduce discontinuities in the rendering equation and therefore need special care. You can try to modify the `spot.cpp` file to expose those in the `traverse()` method at your own risk ;)

**Comment by back2yes** [comment]
So, how should one create a `mi.Ray3f` object with pytorch tensor?

**Comment by back2yes** [comment]
Made a work-around wrapper that takes argument types as inputs as well, so that I can cast pytorch tensors into types other than TensorXf, e.g. Point3f/Array3f.

```
import warnings
from functools import wraps as _wraps
from typing import Callable, Sequence as _Sequence, Mapping as _Mapping

import drjit as _dr


def work_around_ad_wrap(source: str, target: str, arg_types: _Sequence):
    if not 'drjit' in [source, target]:
        raise TypeError('my_wrap(): invalid combination of frameworks, '
                        'expected one to be drjit!', source, target)

    def wrapper(func: Callable):
        @_wraps(func)
        def f(*args, **kwargs):
            assert len(args) == len(arg_types), "Remember to specify `arg_types` for each parameter!"

            if 'torch' in [source, target]:
                import torch as _torch

            # Return whether the input argument is a PyTorch tensor
            def is_torch_tensor(a):
                return getattr(a, '__module__', None) == 'torch' \
                    and type(a).__name__ == 'Tensor'

            # Construct the full tuple of positional arguments
            args = _dr.detail.get_args_values(func, *args, **kwargs)

            # Casting routing from Dr.Jit tensors to PyTorch tensors
            def drjit_to_torch(a, enable_grad=False):
                if isinstance(a, _Sequence) and not isinstance(a, str):
                    return tuple(drjit_to_torch(b, enable_grad) for b in a)
                elif isinstance(a, _Mapping):
                    return {k: drjit_to_torch(v, enable_grad) for k, v in a.items()}
                # elif _dr.is_array_v(a) and _dr.is_tensor_v(a):
                elif _dr.is_array_v(a) and _dr.is_tensor_v(a):
                    b = a.torch()
                    b.requires_grad = _dr.grad_enabled(a) or (enable_grad and _dr.is_diff_v(a))
                    return b
                elif _dr.is_diff_v(a) and a.IsFloat:
                    warnings.warn("my_wrap(): differential input arguments should be Dr.Jit tensor!")
                    b = a.torch()
                    b.requires_grad = _dr.grad_enabled(a) or (enable_grad and _dr.is_diff_v(a))
                    return b
                    # raise TypeError("my_wrap(): differential input arguments "
                    #                 "should be Dr.Jit tensor!")

                else:
                    return a

            # Casting routing from PyTorch tensors to Dr.Jit tensors
            def torch_to_drjit(a, arg_types):
                if isinstance(a, _Sequence) and not isinstance(a, str):
                    return tuple(torch_to_drjit(b, t) for (b, t) in zip(a, arg_types))
                elif isinstance(a, _Mapping):
                    return {k: torch_to_drjit(v, arg_types[k]) for (k, v) in a.items()}
                elif is_torch_tensor(a):
                    dtype_str = {
                        _torch.float: 'TensorXf',
                        _torch.float32: 'TensorXf',
                        _torch.float64: 'TensorXf64',
                        _torch.int32: 'TensorXi',
                        _torch.int: 'TensorXi',
                        _torch.int64: 'TensorXi64',
                        _torch.long: 'TensorXi64',
                    }[a.dtype]

                    device = 'cuda' if a.is_cuda else 'llvm'
                    m = getattr(getattr(_dr, device), 'ad')

                    return _dr.unravel(arg_types, getattr(m, dtype_str)(a))
                else:
                    return a

            # Ensure tensors in `a` have same shape as tensors in `b` (handles dim==0 case)
            def torch_ensure_shape(a, b):
                if isinstance(a, _Sequence) and not isinstance(a, str):
                    return tuple(torch_ensure_shape(a[i], b[i]) for i in range(len(a)))
                elif isinstance(a, _Mapping):
                    return {k: torch_ensure_shape(v, b[k]) for k, v in a.items()}
                elif is_torch_tensor(a):
                    return a.reshape(b.shape)
                else:
                    return a

            if source == 'torch':
                class ToDrJit(_torch.autograd.Function):
                    @staticmethod
                    def forward(ctx, *args):
                        ctx.args = args
                        ctx.args_drjit = torch_to_drjit(args, arg_types)
                        _dr.enable_grad(ctx.args_drjit)
                        res = func(*ctx.args_drjit)
                        ctx.res_drjit = (res,) if not isinstance(res, tuple) else res
                        return drjit_to_torch(res)

                    @staticmethod
                    @_torch.autograd.function.once_differentiable
                    def backward(ctx, *grad_output):
                        _dr.set_grad(ctx.res_drjit, grad_output)
                        _dr.enqueue(_dr.ADMode.Backward, ctx.res_drjit)
                        _dr.traverse(ctx.res_drjit, _dr.ADMode.Backward)
                        args_grad = drjit_to_torch(_dr.grad(ctx.args_drjit))
                        args_grad = torch_ensure_shape(args_grad, ctx.args)
                        del ctx.res_drjit, ctx.args_drjit
                        return args_grad

                return ToDrJit.apply(*args)

            if target == 'torch':
                class ToTorch(_dr.CustomOp):
                    def eval(self, *args):
                        self.args = args
                        self.args_torch = drjit_to_torch(args, enable_grad=True)
                        self.res_torch = func(*self.args_torch)
                        return torch_to_drjit(self.res_torch)

                    def forward(self):
                        raise TypeError("warp_ad(): forward-mode AD is not supported!")

                    def backward(self):
                        grad_out_torch = drjit_to_torch(self.grad_out())
                        grad_out_torch = torch_ensure_shape(grad_out_torch, self.res_torch)

                        def flatten(values):
                            """Flatten structure in a consistent arbitrary order"""
                            result = []

                            def traverse(values):
                                if isinstance(values, _Sequence):
                                    for v in values:
                                        traverse(v)
                                elif isinstance(values, _Mapping):
                                    for _, v in sorted(values.items(), key=lambda item: item[0]):
                                        traverse(v)
                                else:
                                    result.append(values)

                            traverse(values)

                            # Single item should not be wrapped into a list
                            if not isinstance(values, _Sequence) and not isinstance(values, _Mapping):
                                result = result[0]

                            return result

                        _torch.autograd.backward(flatten(self.res_torch), flatten(grad_out_torch))

                        def get_grads(args):
                            if isinstance(args, _Sequence) and not isinstance(args, str):
                                return tuple(get_grads(b) for b in args)
                            elif isinstance(args, _Mapping):
                                return {k: get_grads(v) for k, v in args.items()}
                            elif is_torch_tensor(args):
                                return getattr(args, 'grad', None)
                            else:
                                return None

                        args_grad_torch = get_grads(self.args_torch)
                        args_grad = torch_to_drjit(args_grad_torch)
                        self.set_grad_in('args', args_grad)

                return _dr.custom(ToTorch, args)

        return f

    return wrapper
```

Then we can wrap the Mitsuba function like this:

```
@work_around_ad_wrap(source="torch", target="drjit", arg_types=(mi.Point3f, Array3f))
def torch_render(ray_o, ray_d):
    ray = mi.Ray3f(ray_o, ray_d)
    si = scene.ray_intersect(ray)
    t = si.t
    n = si.n
    return t, n
```
